dll/win32/msafd/misc/sndrcv.c

```diff
@@ -514,44 +589,59 @@ WSPRecvFrom(SOCKET Handle,
                                     APCContext,
                                     IOSB,
                                     IOCTL_AFD_RECV_DATAGRAM,
-                                    &RecvInfo,
-                                    sizeof(RecvInfo),
+                                    RecvInfo,
+                                    sizeof(*RecvInfo),
                                     NULL,
                                     0);
-
-    /* Wait for completion of not overlapped */
-    if (Status == STATUS_PENDING && lpOverlapped == NULL)
+    if (Status == STATUS_PENDING)
     {
-        WaitForSingleObject(SockEvent, INFINITE); // BUGBUG, shouldn wait infinitely for receive...
+        /* Wait for completion if not overlapped */
+        if (!lpOverlapped)
+        {
+            /* FIXME: Shouldn't wait infinitely for receive... */
+            WaitForSingleObject(SockEvent, INFINITE);
```

Comment:
No idea whether this is documented somewhere, but could you try introducing a finite timeout somewhere?
(maybe a `#define RECV_TIMEOUT (3 * 60 * 1000) // 3 minutes` or something similar, and use it here and perhaps in other places where a finite wait should be used instead?)
(Note that I don't know whether there is a registry setting for such timeout value for AFD)

Reply:
This requires more investigation how Windows handles that "timeout" situation and properly terminate the I/O. I think waiting infinitely here is good enough for now.
For the current code base, if we don't wait here until the I/O is completed and early return from the function, it sometimes causes BSOD 0x7E/0x1E because the caller-supplied buffers can be invalidated while the I/O is still in progress.